src/java.base/share/classes/java/io/ByteArrayInputStream.java

```diff
@@ -207,10 +207,21 @@ public int readNBytes(byte[] b, int off, int len) {
     public synchronized long transferTo(OutputStream out) throws IOException {
         int len = count - pos;
         if (len > 0) {
+            byte[] tmp;
+            if (out.getClass().getPackageName().startsWith("java.") &&
+                !(out instanceof FilterOutputStream))
+                tmp = null;
+            else
+                tmp = new byte[Integer.min(len, MAX_TRANSFER_SIZE)];
+
             int nwritten = 0;
             while (nwritten < len) {
                 int nbyte = Integer.min(len - nwritten, MAX_TRANSFER_SIZE);
-                out.write(buf, pos, nbyte);
+                if (tmp != null) {
+                    System.arraycopy(buf, pos, tmp, 0, nbyte);
```

Comment:
I assume the overall performance of transferTo will be faster if we use System.arraycopy *only once* in line 215 to create a safe copy of the *complete* buf instead of calling it multiple times in a loop to create copies *per slice*. In that case we can omit the tmp == null case but simply use tmp = buf, making the code in the loop if-free.

Reply:
There is a tradeoff here between number of invocations of `arraycopy` and amount of memory allocated for `tmp`. (We have seen this before in #14981 which I have allowed to languish.) The allocation limit is `MAX_TRANSFER_SIZE` which is presently 128 kB, so any transfer of size less than this will invoke `arraycopy` only once already.